src/deviation.jl

```diff
@@ -9,13 +9,10 @@ Count the number of indices at which the elements of the arrays
 `a` and `b` are equal.
 """
 function counteq(a::AbstractArray, b::AbstractArray)
-    n = length(a)
-    length(b) == n || throw(DimensionMismatch("Inconsistent lengths."))
+    length(a) == length(b) || throw(DimensionMismatch("Inconsistent lengths."))
     c = 0
-    for i = 1:n
-        @inbounds if a[i] == b[i]
-            c += 1
-        end
+    @inbounds for i = eachindex(a, b)
```

Comment:
This is breaking, as e.g. the following example shows (similar to https://github.com/JuliaStats/StatsBase.jl/pull/790/files#r874144995 without OffsetArrays which was unsupported):
```julia
julia> a = rand(3, 5);

julia> b = view(rand(10, 7), 1:2:10, 1:2:6);

julia> s = 0.0;

julia> for i in eachindex(a, b)
           s += (a[i] != b[i])
       end
ERROR: DimensionMismatch("all inputs to eachindex must have the same axes, got (Base.OneTo(3), Base.OneTo(5)) and (Base.OneTo(5), Base.OneTo(3))")
Stacktrace:
 [1] throw_eachindex_mismatch_indices(::IndexCartesian, ::Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}, ::Vararg{Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}})
   @ Base ./abstractarray.jl:286
 [2] eachindex
   @ ./multidimensional.jl:372 [inlined]
 [3] eachindex(A::Matrix{Float64}, B::SubArray{Float64, 2, Matrix{Float64}, Tuple{StepRange{Int64, Int64}, StepRange{Int64, Int64}}, false})
   @ Base ./abstractarray.jl:328
 [4] top-level scope
   @ ./REPL[9]:1
```
So either we should use `for (ai, bi) in zip(a, b)` or `for (i, j) in zip(eachindex(a), eachindex(b))`. In any case it seems wrong to assume that `a` and `b` have compatible axes and we can use `eachindex(a, b)`.

Reply:
that's really unfortunate.... can we use `firstindex():lastindex()`? `zip` is really slow

Reply:
What exactly do you have in mind? In general, I guess they might be different for `a` and `b` (and I'm not sure if one may assume that the range yields only valid indices).

Reply:
Distances uses a check and otherwise falls back to `zip`: https://github.com/JuliaStats/Distances.jl/blob/91f51b543ea6c54936d3e6183acdf7da50bf1f9e/src/metrics.jl#L251

Reply:
Did you benchmark the `sum(Broadcast.instantiate(Broadcast.broadcasted((ai, bi) -> ai != bi, vec(a), vec(b))))` alternative as well?

Reply:
Timing these, the Broadcast version is fast for big arrays, but slower on small ones. Especially when `vec` causes an allocation: 
```
julia> let x = rand(1000), y = rand(1000) 
         @btime sqL2dist($x, $y)
         @btime sqL2dist_bc($x, $y)
         @btime sum(abs2(ai - bi) for (ai, bi) in zip($x, $y))
       end;
  1.454 μs (0 allocations: 0 bytes)
  450.716 ns (0 allocations: 0 bytes)
  1.462 μs (0 allocations: 0 bytes)

julia> let x = rand(1,5), y = rand(5)
         @btime sqL2dist($x, $y)
         @btime sqL2dist_bc($x, $y)
         @btime sum(abs2(ai - bi) for (ai, bi) in zip($x, $y))
       end;
  7.132 ns (0 allocations: 0 bytes)
  57.334 ns (2 allocations: 80 bytes)
  8.133 ns (0 allocations: 0 bytes)
```

Reply:
The main advantage of the broadcast version is that it uses pairwise summation whereas the other approaches don't.

Reply:
Without the reshape it's within error, I guess. Or you can mess with ReshapedArray, same speed:
```
julia> let x = rand(5,1), y = rand(5)
         @btime sqL2dist($x, $y)
         @btime sqL2dist_bc(Base.ReshapedArray($x, (5,), ()), $y)
         @btime sum(abs2(ai - bi) for (ai, bi) in zip($x, $y))
       end;
  7.132 ns (0 allocations: 0 bytes)
  9.676 ns (0 allocations: 0 bytes)
  8.133 ns (0 allocations: 0 bytes)
```
However, `vec` does not do the same as zip for offset arrays. So my guess is that this broadcast ought to apply only when the axes agree, surely the common case, and fall back to zip? Write your own `_mapreduce` with this non-standard behaviour for this file?

Reply:
Generally, I think the behaviour of `zip(a, b)` is what's implied by the existing code (the current code works eg also for the example above for which `eachindex(a, b)` fails) and what's most consistent with other functions in StatsBase (and similar packages such as Distances, see above).

Reply:
I wish I was surprised that no strange shapes are tested. That, and clarifying the docstring, would be good to do at the same time, IMO.